feature/gnoi/os/tests/osinstall/README.md

```diff
@@ -2,78 +2,96 @@
+
+      b. Transfer the content by issuing gnoi.os.Install rpc with InstallRequest.transfer_content message.
+         * Expect it to return InstallResponse with a TransferProgress status
+         asynchronously at certain intervals.
+         * TODO: When the expected amount of bytes_received is reported by the
+         switch, move to the activation step next.
+
+      c. End the transfer of software by issuing gnoi.os.Install rpc with InstallRequest.TransferEnd message.
+         * Expect the switch to return InstallResponse with a Validated message. The version in the message should be set to the one which was transferred above.
+      
+      d. Activate the software by issuing gnoi.os.Activate rpc.
+         * Set the version field of the ActivateRequest message to be the same as the version specified in the TransferRequest message above.
+         * Set the no_reboot flag to true.
+         * Set the standby_supervisor to FALSE.
+           
+   2. Install and activate the same new software version on standby supervisor:
+      
+      a. Repeat the above process of TransferRequest. This time set the standby_supervisor to TRUE.
+         * Expect the switch to return a InstallResponse with a SyncProgress message. The switch should sync the software image from primary SUP to standby.
+         * Expect the sync to return a value of 100 for percentage_transferred field.
+         * At the end, expect the switch to return InstallResponse with a Validated message. The version in the message should be set to the one which was transferred above.
+           
+   3. Activate the software by issuing gnoi.os.Activate rpc as in the case of primary supervisor.
+      
+      a. Set the version field of the ActivateRequest message to be the same as the version specified in the TransferRequest message above.
+      
+      b. Set the no_reboot flag to true.
+      
+      c. Set the standby_supervisor to TRUE this time.
+      
+   4. Reboot the switch:
+      
+      a. Issue gnoi.system.Reboot as specified in [gNOI-3.1: Complete Chassis Reboot](feature/gnoi/tests/complete_chassis_reboot/complete_chassis_reboot_test.md).
+      
+   5. Verify that the supervisor image has moved to the new image:
+       
+      a. Verify that the supervisor has a valid image by issuing gnoi.os.Verify rpc.
+         * Expect a VerifyResponse with the version field set to the version specified in messages above eventually.
+         * Verify the standby supervisor version.
+         * Expect that the VerifyResponse.verify_standby has the same version in messages above.
+
+               
+* Subtest-2 : Configuration push verification post OS change and check health of the software-module that allows configuration.
+  1. Check the health of the software-module component that allows configuration of the router and verify if it is healthy using the leaf /components/component[**process that handles configuration of the DUT**]/healthz/status/
```

Comment:
IMO, the goal should be to add a, '[smoke test](https://en.wikipedia.org/wiki/Smoke_testing_(software))' to confirm a successful software upgrade occurred.  

Instead of introspection into implementation specific process names, we can accomplish the same goal by pushing a configuration and confirming it is accepted and applied using gNMI.   Coverage of gNOI healthz rpc's should be in other, healthz specific tests IMO.

In other words, remove step 1.    

Step 3 can be changed to verify configuration is the same as what is pushed with gNMI.Get

Remove step 4.

Reply:
i think we should try to use healthz for this test - why wouldn't wouldn't this be the entity library providing a clear "lookup" for this component

Reply:
@dplore 
- Remove step 1:
  - I added this step because it is possible that a config already present in the router may cause config daemon to crash after software upgrade because the new software no longer supports a particular option in the config. So before we go to step2, imo verifying healthz right after OS upgrade is important.
 - Remove step 4:
   - Added step4 to test `config daemon failure reporting` here because it seemed more relevant to the OS install and config compatibility check. If the installed OS is compatible w/ the pushed configuration thats good, but then is the artifacts collection part of this daemon is functioning post upgrade? If not, then the OS is buggy and the OS install test failed.

Reply:
Hi @sachendras 
Re
> I added this step because it is possible that a config already present in the router may cause config daemon to crash after software upgrade because the new software no longer supports a particular option in the config. 

isn't the general assumption that such configuration should be either
a) automatically transformed during the upgrade, or
b) rejected by the node (instead of causing a crash)?
I believe both behaviors are widely implemented.

Indeed a daemon may crash during a reload if there's a SW bug; however, I believe that this logic should be extended to other daemons as well if the objective to ensure overall post-upgrade health of the node. Why limit to a single daemon?
I.e. pre-upgrade the controller should gather health state of all SW components (that report it / must report it), and compare it with the health state post-upgrade.

> If the installed OS is compatible w/ the pushed configuration thats good, but then is the artifacts collection part of this daemon is functioning post upgrade? If not, then the OS is buggy and the OS install test failed.

But this is not specific to an upgrade procedure per se. Was if the artifacts collection part of this daemon wasn't working pre-upgrade either? Wouldn't a better place for a such procedure be in a healthZ specific test as suggested by Darren?

Reply:
```
isn't the general assumption that such configuration should be either
a) automatically transformed during the upgrade, or
b) rejected by the node (instead of causing a crash)?
I believe both behaviors are widely implemented.
```

Sure, but IDK what is the behavior of all implementations. For now, I dont want to enforce a requirement through this test on that particular use case. We will see in future if that needs to change. Either ways, net result for subtest-1 is a failure.

```
Indeed a daemon may crash during a reload if there's a SW bug; however, I believe that this logic should be extended to other daemons as well if the objective to ensure overall post-upgrade health of the node. Why limit to a single daemon?
I.e. pre-upgrade the controller should gather health state of all SW components (that report it / must report it), and compare it with the health state post-upgrade.
```
The subtest says "Configuration push verification post OS change and check health of the software-module that allows configuration". This test here is trying to address a particular DVT found bug whereby post upgrade an implementation had its config daemon crashing. However, it had the base config loaded and working so streaming and regular forwarding was working, just that we couldnt push new config to the DUT. I would like to keep this test focused only to the daemon that manages configuration and addresses the DVT found bug. I dont want to muddy this test by looking for unrelated daemons. I think heathz check would be more relevant for out-of-context checks for all software-modules like you said.

```
But this is not specific to an upgrade procedure per se. What if the artifacts collection part of this daemon wasn't working pre-upgrade either? Wouldn't a better place for such procedure be in a healthZ specific test as suggested by Darren?
```
Sure, we can move this negative test to HealthZ instead.